File: evidence_research/citations.py
from __future__ import annotations

from .models import Claim, Evidence, SourceDocument


def citation_id(claim: Claim) -> str:
    return f"C{claim.id.removeprefix('claim-')[:14]}"


def audit_citations(report: str, claims: list[Claim]) -> dict[str, object]:
    important = [claim for claim in claims if claim.importance != "low"]
    missing = [claim.id for claim in important if f"[{citation_id(claim)}]" not in report]
    coverage = 1.0 if not important else 1 - len(missing) / len(important)
    return {"passed": not missing, "coverage": coverage, "missing_claim_ids": missing}


def evidence_graph_text(
    *, claims: list[Claim], evidence: list[Evidence], sources: list[SourceDocument]
) -> str:
    evidence_by_id = {item.id: item for item in evidence}
    source_by_id = {source.id: source for source in sources}
    lines = []
    for claim in claims:
        lines.append(
            f'<claim id="{claim.id}" status="{claim.status}" confidence="{claim.confidence}">{claim.text}'
        )
        for evidence_id in claim.evidence_ids:
            item = evidence_by_id.get(evidence_id)
            if not item:
                continue
            source = source_by_id.get(item.source_id)
            lines.append(
                f"[{citation_id(claim)}] {item.stance}; {source.canonical_url if source else item.source_id}; quote: {item.quote}"
            )
        lines.append("</claim>")
    return "\n".join(lines)

File: src/citations.ts
import type { Claim, Evidence, SourceDocument } from './types/research';

export function citationId(claim: Claim) {
  return claim.id.replace(/^claim-/, 'C').slice(0, 16);
}

export function renderCitationInstructions({
  claims,
  evidence,
  sources,
}: {
  claims: Claim[];
  evidence: Evidence[];
  sources: SourceDocument[];
}) {
  const evidenceById = new Map(evidence.map(item => [item.id, item]));
  const sourceById = new Map(sources.map(source => [source.id, source]));
  return claims
    .map(claim => {
      const citations = claim.evidenceIds
        .map(id => evidenceById.get(id))
        .filter((item): item is Evidence => Boolean(item))
        .map(item => {
          const source = sourceById.get(item.sourceId);
          return `[${citationId(claim)}] ${item.stance}; ${source?.canonicalUrl ?? item.sourceId}; quote: ${item.quote}`;
        })
        .join('\n');
      return `<claim id="${claim.id}" status="${claim.status}" confidence="${claim.confidence}">${claim.text}\n${citations || 'NO CITATIONS'}</claim>`;
    })
    .join('\n');
}

export function auditCitations(report: string, claims: Claim[]) {
  const importantClaims = claims.filter(claim => claim.importance !== 'low');
  const missing = importantClaims.filter(
    claim => !report.includes(`[${citationId(claim)}]`),
  );
  return {
    passed: missing.length === 0,
    coverage:
      importantClaims.length === 0
        ? 1
        : 1 - missing.length / importantClaims.length,
    missingClaimIds: missing.map(claim => claim.id),
  };
}
